### ISSUE TITLE
ARM: Bus Error using synchronized blocks (1.1.0-beta6)

### ISSUE THREAD

**Issue by smolt** [ldc-developers/ldc#1955]
1.1.0-beta6 build on Raspbian Pi 3 is getting Bus Errors in four dmd-testsuite tests:

- runnable/eh.d
- runnable/test23.d
- runnable/test34.d
- runnable/test4.d

```
LDC - the LLVM D compiler (1.1.0-beta6):
  based on DMD v2.071.2 and LLVM 3.9.1
  built with LDC - the LLVM D compiler (4a9853)
  Default target: armv7l-unknown-linux-gnueabihf
  Host CPU: (unknown)
  http://dlang.org - http://wiki.dlang.org/LDC

  Registered Targets:
    aarch64    - AArch64 (little endian)
    aarch64_be - AArch64 (big endian)
    arm        - ARM
    arm64      - ARM64 (little endian)
    armeb      - ARM (big endian)
    thumb      - Thumb
    thumbeb    - Thumb (big endian)
```

gdb shows all cases happen in code generated for synchronized blocks, something messed up in D_CRITICAL_SECTION, probably the pthread_mutex_t struct is misaligned.

Working back it appears the Bus Errors started in commit 8d06f23 "Allow align(1) for global variables". I'll have to study it, but my guess is that there was an existing alignment problem and this commit just revealed it.

**Comment by smolt** [comment]
Ok, it is a misaligned `pthread_mutext_t` contained in a D_CRITICAL_SECTION.

Just this generates a bus error:
```D
void main()
{
  synchronized
  {
  }
}
```
Reason: ddmd/statement.d for `synchronized` creates byte array used as a D_CRITICAL_SECTION struct in druntime, and byte array gets alignment of 1 now (was 4).  CPUs with strict alignment don't like this.
```D
            /* Generate our own critical section, then rewrite as:
             *  __gshared byte[CriticalSection.sizeof] critsec;
             *  _d_criticalenter(critsec.ptr);
             *  try { body } finally { _d_criticalexit(critsec.ptr); }
             */
```
Not sure of fix yet.  Seems like there could be a `Target.critsecalignment()` similar to `Target.critsecsize()` used to set alignment of byte array.  Temporary fix would be to undo 8d06f23.

**Comment by rainers** [comment]
I suspect alignment to target.ptrsize would be good enough here. There is even a pointer prepended to the D_CRITICAL_SECTION inside the byte array.

**Comment by thewilsonator** [comment]
Indeed, I'm not sure why the extra space for for a pointer is there at all. I suspect that all that will be needed is to change the byte array to a size_t array as the size of a critical section is always a multiple of size_t.

**Comment by rainers** [comment]
> Indeed, I'm not sure why the extra space for for a pointer is there at all.

There is a chain pointer inside D_CRITICAL_SECTION to allow destroying these at program exit (https://github.com/ldc-developers/druntime/blob/ldc/src/rt/critical_.d#L48).

>  I suspect that all that will be needed is to change the byte array of critsec.sizeof to a size_t array of critsec.sizeof/size_t.sizeof as the size of a critical section is always a multiple of size_t.

Setting `tmp.alignment = Target.ptrsize;` should be fine, too.

**Comment by kinke** [comment]
See #1956. There's a bit more to it, as the critical section size was wrong in most cross-compilation cases.

**Comment by kinke** [comment]
@smolt: A list (incl. output) of the lit-tests failing for your ARM platform would be nice too. Glad you're back. :)

**Comment by kinke** [comment]
Fixed by #1956 and will be in 1.1 final.
